Updated templates to version 5.3 for Deep Security 10.1

trend-scottb opened the pull request.

santiagocardenas commented:
When we merge, I will run a quick beautification on the templates that I need your help in validating to ensure no functionality changes.

santiagocardenas commented on templates/Common/DSMELB.template:
```diff
@@ -1,260 +1,118 @@
+      - Internet-facing
+      - Internal
+    Default: Internet-facing
+  CfnUrlPrefix:
```
We don't recommend using a parameter like this, but instead using QSS3BucketName and QSS3KeyPrefix and dynamically building a virtual-hosted S3 url from it via Fn::Sub similar to https://github.com/aws-quickstart/quickstart-linux-bastion/blob/develop/templates/linux-bastion-master.template#L310

trend-scottb commented:
When I get the merge alert I can perform sanity checks on the templates.

santiagocardenas commented on templates/quickstart/TMQuickStartBYOL.template:
```diff
@@ -0,0 +1,454 @@
+      can include numbers, lowercase letters, uppercase letters, and hyphens (-).
+      It cannot start or end with a hyphen (-).
+    Type: String
+  QSS3KeyPrefix:
```
This whole parameter should look like this (AllowedPattern, ConstraintDescription, Description), but with a Default value of "trendmicro/deepsecurity/latest/": https://github.com/aws-quickstart/quickstart-linux-bastion/blob/develop/templates/linux-bastion-master.template#L225

santiagocardenas replied:
QSS3KeyPrefix has not been corrected as per my comment above.

santiagocardenas commented on templates/quickstart/TMQuickStartBYOL.template:
```diff
@@ -0,0 +1,454 @@
+    ConstraintDescription: Quick Start bucket name can include numbers, lowercase
+      letters, uppercase letters, periods (.), and hyphens (-). It cannot start or
+      end with a hyphen (-).
+    Default: trend-micro-quick-start
```
Default value must be "quickstart-reference"

santiagocardenas commented on templates/quickstart/TMQuickStart.template:
```diff
@@ -0,0 +1,419 @@
+      can include numbers, lowercase letters, uppercase letters, and hyphens (-).
+      It cannot start or end with a hyphen (-).
+    Type: String
+  QSS3KeyPrefix:
```
This whole parameter should look like this (AllowedPattern, ConstraintDescription, Description), but with a Default value of "trendmicro/deepsecurity/latest/": https://github.com/aws-quickstart/quickstart-linux-bastion/blob/develop/templates/linux-bastion-master.template#L225

santiagocardenas replied:
QSS3KeyPrefix has not been corrected yet as per my comment above.

trend-scottb replied:
Thought you were referring to the default value.

The QSS3KeyPrefix in the attached template allows forward slashes at the start or end of the prefix, which i think will crash the template if entered because they are automatically appended to each end. This version (that I think I got from an earlier version of the best practices) doesn't allow that which i think is less error prone.

I also would prefer not expecting the trailing forward slash to be part of the parameter as it makes the template code that uses the parameter less readable.

santiagocardenas replied:
We are following these guidelines across all Quick Starts are they are actual key prefixes. The trailing slash is no longer appended automatically anymore (anywhere) and proper key prefixes (like in the SDKs and CLI) will include the slash. Key prefixes are not directories though we illustrate them as if they were and it's best not to assume characters being added. Please take a look at the example above from Linux bastion as it complies and resembles key prefixes elsewhere.

santiagocardenas replied:
One more thing that is of importance. By having a complete key prefix that includes the slash, it allows for empty key prefixes which are acceptable in S3 buckets. That is, if someone takes the Quick Start and customizes it for their use, they can copy all the objects to the root of their own bucket, which effectively means QSS3KeyPrefix = ""

santiagocardenas commented on templates/quickstart/TMQuickStart.template:
```diff
@@ -0,0 +1,419 @@
+    ConstraintDescription: Quick Start bucket name can include numbers, lowercase
+      letters, uppercase letters, periods (.), and hyphens (-). It cannot start or
+      end with a hyphen (-).
+    Default: trend-micro-quick-start
```
Default value must be "quickstart-reference"

santiagocardenas commented on templates/quickstart/TMQuickStart.template:
```diff
@@ -0,0 +1,419 @@
+      are automatically appended.
+    Type: String
+Mappings:
+  AWSInfoRegionMap:
```
This is no longer used in the Quick Start standards. The URL is dynamically built using Fn::Sub. See other comment around this.

santiagocardenas commented on templates/quickstart/TMQuickStart.template:
```diff
@@ -0,0 +1,419 @@
+    Condition: PerHostSupportedRegion
+    Properties:
+      TemplateURL:
+        !Sub
```
See improved way of doing this: https://github.com/aws-quickstart/quickstart-linux-bastion/blob/develop/templates/linux-bastion-master.template#L261

santiagocardenas commented on templates/quickstart/TMQuickStart.template:
```diff
@@ -0,0 +1,419 @@
+        DBISubnet2: !Ref DatabaseSubnet2
+        DSIPLicense: PerHost
+        DBPMultiAZ: 'true'
+        CfnUrlPrefix:
```
Pass in QSS3BucketName and QSS3KeyPrefix instead.

santiagocardenas commented:
Looks like you have a total of 14 templates. Please add the one of each of the following IDs to each template (1 unique per template). Don't worry about left-over IDs... I provisioned extra ones:
(qs-1ngr590i4)
(qs-1ngr590i9)
(qs-1ngr590ie)
(qs-1ngr590ij)
(qs-1ngr590io)
(qs-1ngr590iu)
(qs-1ngr590j4)
(qs-1ngr590j9)
(qs-1ngr590je)
(qs-1ngr590jj)
(qs-1ngr590jo)
(qs-1ngr590jt)
(qs-1ngr590k4)
(qs-1ngr590k9)
(qs-1ngr590ke)
(qs-1ngr590kj)
(qs-1ngr590ko)
(qs-1ngr590kt)
(qs-1ngr590l3)
(qs-1ngr590l9)

More info: https://aws-quickstart.github.io/qs-ids.html

santiagocardenas commented on scripts/add-aws-account-with-instance-role.sh:
```diff
@@ -1,27 +1,25 @@
 #!/bin/bash
```
We recommend to have #!/bin/bash -e at the top of the bootstrapping scripts (but not directly in the UserData block) so that if something fails, you can trigger and exit and signal a failure to the CloudFormation stack, instead of waiting until timeout.

trend-scottb replied:
Some of the scripts are doing non critical setup and are allowed to fail silently. Would like to submit the scripts as is and revisit this on the next update after I've had time to do more testing on the failure paths.

santiagocardenas replied:
Sounds good.

santiagocardenas commented on templates/quickstart/TMQuickStartBYOL.template:
```diff
@@ -0,0 +1,454 @@
+    Properties:
+      TemplateURL:
+        !Sub
+        - ${QuickStartS3URL}/${QSS3BucketName}/${QSS3KeyPrefix}/templates/marketplace/MasterMP.template
```
paths must match case... I noticed that the directories are Marketplace, etc. We prefer all lowercase, but it means directories must be renamed.

trend-scottb replied:
The case in my local repo is lower case. I'm fine doing whatever we need to do to change this to be consistent.

santiagocardenas commented:
@trend-scottb 
Please take a look at the comments above and I think we'll be closer to merging to the feature branch... from there I'll run a beautifier and iterate a bit more to ensure we have all the benefits of portability from Quick Starts.

trend-scottb commented:
@santiagocardenas 
Sounds good, thanks for the review. I'll start working on those changes.

trend-scottb commented:
Updated code to address the comments in the review, let me know if there's something I've missed.

santiagocardenas commented on templates/common/security-groups/DSMSGIngressRules.template:
```diff
@@ -0,0 +1,179 @@
+      IpProtocol: tcp
+      FromPort: '4122'
+      ToPort: '4122'
+      CidrIp: 10.0.0.0/8
```
Would it be possible to not hardcode these specific CIDR blocks in the ingress rules so that the customer has a choice on subnet sizing? Take a look at the [VPC Quick Start](https://github.com/aws-quickstart/quickstart-aws-vpc) to see how we set some recommended defaults on CIDR blocks, but allow changing based on customer preference.

trend-scottb replied:
We deliberately decided not to expose the CIDR ranges to the customer as most users will be confused by what to enter for the parameter. The expectation is the customer will have private instances located in the DSM VPC (across all subnets) that need to connect with the DSM and these rules allow those connections regardless of what the CIDR range of the VPC has.

santiagocardenas replied:
I see. Would be something then to consider based on customer feedback on flexibility. Good for now though.

santiagocardenas commented on templates/Common/sps.template:
```diff
@@ -1,268 +1,170 @@
+          - - |
+              #/bin/bash
+            - 'chpasswd <<<"admin:'
+            - !Ref SPSWebAdminPassword
```
This is not good. Anyone in that logs on to the instance or can describe the user data can get the password since User Data is not encrypted.

trend-scottb replied:
This is an old template that isn't used in the quick start, will remove it.

santiagocardenas replied:
Can you please do a cleanup of any other assets that are no longer used?

trend-scottb replied:
Will do

santiagocardenas commented:
@trend-scottb Great progress on the changes. I added a few comments.

santiagocardenas commented:
Also, I would recommend switch all file names and directories to lower-case. They become easier to type and less error prone given that S3 object keys are case sensitive and we have human-readable objects.

trend-scottb commented:
Changed the QSS3KeyPrefix as per your comments, removed some all un-referenced files and changed the templates to be lower case. It looks like that last change didn't register as a file move in git but as a delete and add. Apologies there, I think that was everything that was outstanding.
